### PR TITLE
CBL-59: Supporting date parsing functions

### PR DESCRIPTION
**Pull request by callumbirks**
Rebased feature/parse_date_parts onto master.

### CODE REVIEW

**Comment by callumbirks** [comment]
The functions exposed / added here are required to support the date/time functions being added to Core / N1QL

**Comment by callumbirks** [comment]
Seems like some functions from the PRD aren't present in any existing branches. (Specifically, `str_to_tz`, and other tz functions). I think I'll need to add a new dependency for that (`date/tz`). We already have `date/date.h` as a Fleece dependency, we just need the timezone extensions.

File: Fleece/Support/ParseDate.cc [hunk 1]
```diff
@@ -329,124 +309,294 @@ static void inject_local_tz(DateTime* p)
...
+                  [](unsigned char c) { return std::tolower(c); });
+        const auto entry = dateComponentMap.find(componentStr);
+        if ( entry == dateComponentMap.end() ) { return kDateComponentInvalid; }
+
+        return entry->second;
     }
 
-    slice FormatISO8601Date(char buf[], int64_t time, bool asUTC) {
-        if (time == kInvalidDate) {
+    slice FormatISO8601Date(char buf[], int64_t timestamp, bool asUTC, const DateTime* format) {
```
Comment (jianminzhao) [inline Fleece/Support/ParseDate.cc:473]: Why do we need format? Does it mean that the caller may want to discard the YMD or HMS ?

**Comment by jianminzhao** [comment]
Please add some tests

**Comment by callumbirks** [comment]
> Please add some tests

There are lots of tests in the accompanying LiteCore PR (blocked on this one), which test these functions, but through the interface of N1QL

File: Fleece/Support/ParseDate.cc (hunk 1 above)
Comment (callumbirks) [reply Fleece/Support/ParseDate.cc:473]: Yeah, per the PRD, we must support format other than ISO8601. We provide this in N1QL by allowing a DateTime parameter where the caller can pass a DateTime that looks like the format they want. i.e. if they pass "11:11:11" we would return them a datetime that only has HMS

**Comment by callumbirks** [comment]
@jianminzhao There is a test already (`SupportTests.cc:334`) which converts timestamp to string, and back again. That process involves most/all of the functions added/modified here

**Review by pasin — APPROVED** [review]
From @callumbirks, we have tests that well cover the all the parser logic in this PR.

**Review by jianminzhao — APPROVED** [review]

**Pull request merged**